src/IROperator.cpp

```diff
@@ -242,7 +360,25 @@ bool is_positive_const(const Expr &e) {
         return f->value > 0.0f;
     }
     if (const Cast *c = e.as<Cast>()) {
-        return is_positive_const(c->value);
+        Type to = c->type;
+        if (!to.is_int_or_uint() ||
+            lossless_cast(to, c->value).defined()) {
```

Comment:
lossless_cast is quite expensive, and is_positive_const is used in places where it needs to be very cheap. I think we might need to come up with an alternative strategy for handling cast chains. Maybe this condition should just be to.can_represent(c->value.type()) to check for widening casts.

Reply:
Are these expectations of "expected to be cheap", "known to be expensive" etc documented for each of these?

Reply:
No, the responsibilities of these things tends to creep over time (e.g. lossless_cast is becoming more and more heavy-duty), so I'm trying to push back. I just know in the past I've called is_positive_const assuming it'll be cheap. The simplifier used to be littered with it, but that was all stripped out. Now it seems to be used for special cases in bounds, is_monotonic, the solver, and some more bounds stuff in vectorization. I *think* in those contexts the simplifier has already run. If that's the case we should just enshrine that in the comments on is_positive_const (i.e. that it does not do constant folding).